File: OpenGUI/Assets/Editor/OpenGUI/OGRootInspector.js
#pragma strict

@CustomEditor ( OGRoot )
@InitializeOnLoad
public class OGRootInspector extends Editor {
	private static function OGRootInspector () {
		OGRoot.EditorSelectWidget = function ( w : OGWidget, additive : boolean ) {
			if ( !w ) {
				Selection.activeObject = null;

			} else if ( additive ) {
				var list : List.< UnityEngine.Object > = new List.< UnityEngine.Object > ( Selection.objects );
				var remove : boolean = false;

				for ( var i : int = 0; i < list.Count; i++ ) {
					if ( list[i] as GameObject == w.gameObject ) {
						remove = true;
						break;
					}
				}
				
				if ( remove ) {
					list.Remove ( w.gameObject );
				} else {
					list.Add ( w.gameObject );
				}

				Selection.objects = list.ToArray ();
			
			} else {
				Selection.activeObject = w.gameObject;
			
			}
		};
	}
	
	override function OnInspectorGUI () {
		var root : OGRoot = target as OGRoot;
	
		if ( OGRoot.EditorSelectWidget == null ) {
			OGRoot.EditorSelectWidget = function ( w : OGWidget ) {
				Selection.activeObject = w.gameObject;
			};
		}
		
		if ( !root ) { return; }
		
		DrawDefaultInspector ();

		EditorGUILayout.Space ();

		GUI.backgroundColor = Color.green;
		if ( GUILayout.Button ( "Update", GUILayout.Height ( 30 ) ) ) {
			root.SetDirty ();
		}
		GUI.backgroundColor = Color.white;

	}
}

File: OpenGUI/Assets/Editor/OpenGUI/OGRootInspector.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using UnityEditor;

[CustomEditor (typeof(OGRoot))]
[InitializeOnLoad]
public class OGRootInspector : Editor {
	static OGRootInspector () {
		OGRoot.EditorSelectWidget = (w, additive) => {
			if ( !w ) {
				Selection.activeObject = null;

			} else if ( additive ) {
				List< UnityEngine.Object > list = new List< UnityEngine.Object > ( Selection.objects );
				bool remove = false;

				for ( int i = 0; i < list.Count; i++ ) {
					if ( list[i] as GameObject == w.gameObject ) {
						remove = true;
						break;
					}
				}
				
				if ( remove ) {
					list.Remove ( w.gameObject );
				} else {
					list.Add ( w.gameObject );
				}

				Selection.objects = list.ToArray ();
			
			} else {
				Selection.activeObject = w.gameObject;
			
			}
		};
	}
	
	public static void SelectActiveObject ( OGWidget w, bool b = false )
	{
		Selection.activeObject = w.gameObject;
	}

	override public void OnInspectorGUI () {
		OGRoot root = (OGRoot) target;
	
		if ( OGRoot.EditorSelectWidget == null ) {
			OGRoot.EditorSelectWidget = SelectActiveObject;
		}
		
		if ( !root ) { return; }
		
		DrawDefaultInspector ();
	}
}
